BPF: Disable atomic CAS

vadorovsky opened the pull request:
Enabling CAS for BPF targets (https://github.com/rust-lang/rust/pull/105708) breaks the build of core library.
The failure occurs both when building rustc for BPF targets and when
building crates for BPF targets with the current nightly.

The LLVM BPF backend does not correctly lower all `atomicrmw` operations
and crashes for unsupported ones.

Before we can enable CAS for BPF in Rust, we need to fix the LLVM BPF
backend first.

Fixes #106795

Signed-off-by: Michal Rostecki <vadorovsky@gmail.com>

vadorovsky commented:
/cc @nagisa @tomerze @alessandrod

alessandrod commented:
There are at least two separate issues here I think:

- The `atomic_int` macro is defined only for `[cfg(target_has_atomic_load_store = "8")]`. That should probably be changed to check for all the `_load_store` sizes.
- The LLVM BPF backend does not correctly lower all `atomicrmw` ops, and will crash for unsupported ones, so `atomic_cas` should be reverted back to false. The `min_atomic_width` definition should probably stay tho.

cjgillot commented:
r? @nagisa

tomerze commented:
We should probably fix the first issue regardless, for people who might be creating a custom target json.

alessandrod commented:
> We should probably fix the first issue regardless, for people who might be creating a custom target json.

Yes, I think we should just revert `atomic_cas` back to false here, then fix the `atomic_int` macro definition in a separate PR.

vadorovsky commented:
> > There are at least two separate issues here I think:
> > 
> >     * The `atomic_int` macro is defined only for `[cfg(target_has_atomic_load_store = "8")]`. That should probably be changed to check for all the `_load_store` sizes.
> > 
> >     * The LLVM BPF backend does not correctly lower all `atomicrmw` ops, and will crash for unsupported ones, so `atomic_cas` should be reverted back to false. The `min_atomic_width` definition should probably stay tho.
> > 
> > 
> > The bpfel-* targets are completely broken for non trivial programs (pretty much all aya programs and CI are broken with recent nightly), so I'd like to get this merged asap.
> 
> We should probably fix the first issue regardless, for people who might be creating a custom target json.

I already prepared a patch for the first issue, I will submit it soon.

vadorovsky committed:
BPF: Disable atomic CAS

Enabling CAS for BPF targets (#105708) breaks the build of core library.
The failure occurs both when building rustc for BPF targets and when
building crates for BPF targets with the current nightly.

The LLVM BPF backend does not correctly lower all `atomicrmw` operations
and crashes for unsupported ones.

Before we can enable CAS for BPF in Rust, we need to fix the LLVM BPF
backend first.

Fixes #106795

Signed-off-by: Michal Rostecki <vadorovsky@gmail.com>

vadorovsky commented:
@alessandrod According to your suggestion, I'm changing here only `atomic_cas` to `false`. Therefore, it's not a simple revert anymore, so I added a more elaborate commit message.

Please note that this PR will need also #106856 to fully fix BPF. Since we keep `min_atomic_width` untouched (`64`), the problem with `#[cfg(target_has_atomic = "8")]` still remains. So we need to merge both of them

vadorovsky commented:
@nagisa ping

vadorovsky commented:
/cc @bjorn3

bjorn3 commented:
This change makes sense to me. It was already disabled up until very recently and is clearly broken.

@bors r+

bors commented:
:pushpin: Commit 651e8734620a07d25acab4a594e23e33b4d91f50 has been approved by `bjorn3`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=bjorn3 651e8734620a07d25acab4a594e23e33b4d91f50 -->
<!-- homu: {"type":"Approved","sha":"651e8734620a07d25acab4a594e23e33b4d91f50","approver":"bjorn3","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

taiki-e commented:
This does not appear to be correct because min-atomic-width is still set.

Below is an example of disabling atomic-cas as this PR does in the bpf target-spec for the toolchain affected by the regression, but it fails with the same error as #106795.

```console
$ cargo new --lib a
$ cd a
$ rustc -Z unstable-options --target bpfel-unknown-none --print target-spec-json | grep -v is-builtin > bpfel-unknown-none.json
$ sed -i 's/"arch": "bpf",/"arch": "bpf","atomic-cas": false,/' bpfel-unknown-none.json
$ echo '#![no_std]' > src/lib.rs
$ cargo build -Z build-std=core --target bpfel-unknown-none.json
error: cannot find macro `atomic_int` in this scope
    --> /Users/taiki/.rustup/toolchains/nightly-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/sync/atomic.rs:2857:1
     |
2857 | atomic_int! {
     | ^^^^^^^^^^

...
```

I would suggest simply reverting https://github.com/rust-lang/rust/pull/105708 or commenting out `min_atomic_width: Some(64)` in `bpf_base.rs` for now, and handle small width atomics in #106856.

bjorn3 commented:
Did you try this PR merged with https://github.com/rust-lang/rust/pull/106856?

bjorn3 commented:
https://github.com/rust-lang/rust/pull/106856 should be enough to fix that error, right?

taiki-e commented:
Oh, never mind, I misunderstood that this was a PR to fix https://github.com/rust-lang/rust/issues/106845.

bjorn3 commented:
I believe that PR was split out of this one. I did notice that this PR still has "Fixes https://github.com/rust-lang/rust/issues/106795" in the description, so the issue will need to be re-opened once this PR lands.

vadorovsky commented:
Sorry for the confusion with issues! Yes, the other PR got split from this one, since we created a separate issue for fixing the annotations - #106845

When in fact, this PR is a partial fix for #106795 and the other PR fixes both issues ultimately.
